ai/prompts/PYRE_ANALYSIS.md

```diff
@@ -0,0 +1,254 @@
+
+### Attempt 2: AST Script (Practical alternative)
+```python
+# plans/fix_dataset_file_id_invalidation/analyze_attribute_modifications.py
+import ast
+
+class MethodAnalyzer(ast.NodeVisitor):
+    def visit_Assign(self, node):
+        # Track all assignments to _dataset_id
+        if isinstance(target, ast.Attribute) and target.attr == '_dataset_id':
+            # Record method name and line number
+            ...
+```
+
+**Result**: Completes in < 1 second, found all 5 buggy methods.
+
+**Lesson**: For custom analysis patterns, AST scripts are often more practical than pyre queries.
+
+## Troubleshooting
+
+### Pyre Times Out
+
+**Symptom**: Hangs at "Processed 1631 of 1709 functions" for 5+ minutes.
+
+**Solutions**:
+1. Use incremental mode after first full run: `pyre incremental`
+2. Analyze smaller subsets: `pyre check graphistry/PlotterBase.py`
+3. Use AST scripts for custom analysis instead
+4. Increase timeout: `pyre check --timeout 600`
+
+### Library Incompatibility
+
+**Symptom**: `GLIBC_2.33 not found` or `GLIBCXX_3.4.29 not found`
+
+**Solutions**:
+1. Use Docker with python:3.12-slim image (has newer libraries)
+2. Upgrade system libraries (may not be feasible)
+3. Use alternative analysis tools (AST scripts, mypy)
+
+### Out of Memory
+
+**Symptom**: Pyre crashes or system becomes unresponsive.
+
+**Solutions**:
+1. Close other applications
+2. Analyze smaller file subsets
+3. Use pyre in Docker with memory limits: `docker run --memory=4g ...`
+
+## Integration with Development Workflow
+
+### Pre-commit Type Checking
+
+```bash
+# Run mypy instead of pyre for faster type checking
+./bin/mypy.sh
+
+# Or in Docker
+cd docker && WITH_BUILD=0 WITH_TEST=0 WITH_LINT=0 WITH_TYPECHECK=1 ./test-cpu-local.sh
+```
+
+**Note**: PyGraphistry uses mypy for regular type checking. Pyre is for advanced analysis only.
+
+### Refactoring Workflow
+
+1. **Identify**: Use pyre/grep to find all occurrences
+2. **Analyze**: Use AST scripts for custom patterns
+3. **Refactor**: Make changes with type awareness
+4. **Validate**: Run mypy + tests to verify
+
+## Pyre vs Mypy
+
+| Feature | Pyre | Mypy |
+|---------|------|------|
+| Type checking | ✅ Yes | ✅ Yes |
+| Speed | Slower (5-10min) | Faster (< 1min) |
+| Query API | ✅ Yes | ❌ No |
+| Call graphs | ✅ Yes | ❌ No |
+| Find references | ✅ Yes | ❌ No |
+| PyGraphistry default | ❌ No | ✅ Yes |
+
+**Recommendation**: Use mypy for regular type checking, pyre for advanced refactoring analysis.
+
+## Future Work
+
+- **Performance**: Investigate pyre incremental mode for faster analysis
+- **Integration**: Add pyre query scripts for common refactoring tasks
+- **Documentation**: Document useful query patterns as they're discovered
+- **CI/CD**: Consider pyre for static analysis in CI (if performance improves)
+
+## Resources
+
+- [Pyre Documentation](https://pyre-check.org/)
+- [Pyre Query API](https://pyre-check.org/docs/querying-pyre/)
+- PyGraphistry AST Scripts: `plans/*/analyze_*.py`
```

Comment:
./plans/ is .gitignored, move analyze_*.py reference script into ai/ folder somewhere (assets?)